File: web/src/app/interfaces/entities/supplier.ts
import {Contact} from './contact';
import {ResourceRelations} from '../resource';
import {Address} from './address';

export interface Supplier {
  id: string | number;
  name: string;
}

export type SupplierCreateRequest = Omit<Supplier, 'id'>;

export type SupplierUpdateRequest = Supplier;

export interface SupplierWithContacts extends Supplier {
  contacts: Contact[];
}

export interface SupplierDetail extends SupplierWithContacts {
  addresses: Address[];
}


export type SupplierRelations = ResourceRelations<['contacts', 'purchaseOrders']>

File: server/src/main/java/com/warehouse/server/entities/Supplier.java
package com.warehouse.server.entities;

import com.fasterxml.jackson.annotation.JsonManagedReference;
import jakarta.persistence.*;
import jakarta.validation.constraints.NotNull;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.springframework.data.rest.core.config.Projection;

import java.util.ArrayList;
import java.util.List;

@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
@Entity
@Table(name = "suppliers")
public class Supplier {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @NotNull
    private String name;

    @OneToMany(mappedBy = "supplier", cascade = CascadeType.ALL, orphanRemoval = true)
    @Builder.Default
    @JsonManagedReference
    private List<Address> addresses = new ArrayList<>();

    @OneToMany(mappedBy = "supplier", cascade = CascadeType.ALL, orphanRemoval = true)
    @Builder.Default
    @JsonManagedReference
    private List<Contact> contacts = new ArrayList<>();

    @OneToMany(mappedBy = "supplier", cascade = CascadeType.ALL, orphanRemoval = true)
    @Builder.Default
    private List<PurchaseOrder> purchaseOrders = new ArrayList<>();

    @Projection(name = "detail", types = {Supplier.class})
    public interface SupplierDetailProjection {
        Long getId();

        String getName();

        List<Address> getAddresses();

        List<Contact> getContacts();
    }
}
